### ISSUE TITLE
Gradle copy task does not preserve file modification times

### ISSUE THREAD

**Issue by jochenberger** [gradle/gradle#1252]
Copied from https://issues.gradle.org/browse/GRADLE-2698

The Gradle copy task does not preserve file modification times even when not changing the file contents.

There should at least be an option to preserve these times.

A sample build file that shows the issue:
```groovy
defaultTasks 'clean', 'copy'

def sourceFile = file("src/foo")

task createFile {
  doLast {
    sourceFile.parentFile.mkdirs() 
    sourceFile.text = "test" 
    sleep 1000 // if the file system has only second precision
  }
}

task copy(type:Copy, dependsOn: createFile){
  from "src"
  into "target"
  doLast {
    assert sourceFile.lastModified() == file('target/foo').lastModified()
  }
}

task clean {
  doLast {
    delete "src", "target"
  }
}
```

A possible solution would be extending `AbstractFileTreeElement.copyTo(File)` by `target.setLastModified(getLastModified());`, but that would make it the default behavior. I'm not sure if that's a good idea.

**Comment by jochenberger** [comment]
Of course, this example does not make a lot of sense. In my use case, I'd like to preserve the modification times when extracting files from `zipTree`.

**Comment by Horcrux7** [comment]
Is there any workarounds?

**Comment by w25r** [comment]
@Horcrux7 a workaround is to use [Ant's copy task](https://ant.apache.org/manual/Tasks/copy.html), which can be called directly from Gradle and has a `preservelastmodified ` option.

**Comment by mkalb** [comment]
I have the same issue with the 'Sync' task.

1. It does not preserve the original modification time.
2. If the task is not 'up-to-date', then it changes **all** outputfiles to the current time.

**Comment by kno10** [comment]
Workaround (preserve timestamps) example:
```
task collectDependencies(type: Copy) {
        from configurations.runtime
        into "$project.buildDir/libs"
        // Try to preserve timestamps:
        def copyDetails = []
        eachFile { copyDetails << it }
        doLast {
                copyDetails.each { FileCopyDetails details ->
                        def target = new File(destinationDir, details.path)
                        if(target.exists()) { target.setLastModified(details.lastModified) }
                }
        }
}
```

**Comment by Mahoney** [comment]
Could I get a feel for whether a patch to do this would be accepted?

When building a docker container I like to `COPY` external dependencies in one step, then binaries generated by the build in a second step, so that in the normal case where the external dependencies haven't changed the existing cached layer is used; but the change to the last modified breaks that because the files are always different, so the cache never works.

Possible approaches:
* Add `setPreserveLastModified(Boolean preserveLastModified)` to `CopyProcessingSpec`
  Default false, setting to true would preserve the last modified timestamp
  ```kotlin
  copy {
    into("external")
    from(externalDependencies)
    preserveLastModified = true
  }
  ```
* Add `setLastModified(long time)` to `CopyProcessingSpec`
  Allows setting the last modified time of the destination file arbitrarily; `eachFile` would then allow retrieving the original file's lastModified and calling this method to maintain it:
  ```kotlin
  copy {
    into("external")
    from(externalDependencies)
    eachFile {
      lastModified = file.lastModified
    }
  }
  ```
* Add `eachFilePostCopy` to `CopyProcessingSpec`, with references to both the original and the copied file, allowing arbitrary mutation of the File _after_ it has been created in the new location with reference to its properties in the old location.
  ```kotlin
  copy {
    into("external")
    from(externalDependencies)
    eachFilePostCopy {
      destination.file.lastModified = source.file.lastModified
    }
  }
  ```
Any or all of these would allow solving the problem in the context of the `CopySpec`, in increasingly general ways. Perhaps even allow all of them?

**Comment by Horcrux7** [comment]
As an enduser I would prefer `preserveLastModified = true`. This is the most understandable if someone later reads the Gradle script.

**Comment by kno10** [comment]
In my opinion, `preserveLastModified` should be standard, as a "copy" does not modify file contents.
Allowing users to choose a different behavior is, of course, good. If you want to update the file stamps, a simple `eachFile { touch }` would be enough (c.f. unix `touch` command).

**Comment by Mahoney** [comment]
As per my edit above, I was totally wrong about Docker's layer caching behaviour; Docker caches the layer regardless of last modified times on files if their contents are identical. I have to say this substantially reduces my motivation to supply a PR for this, I'm afraid, as my use case is gone.

**Comment by cocorossello** [comment]
It also happens with the build cache in windows. See https://youtrack.jetbrains.com/issue/IDEA-260587

**Comment by kkurczewski** [comment]
> Docker caches the layer regardless of last modified times on files if their contents are identical.

For people coming from JVM land, Docker may don't care about timestamps (I didn't confirmed this) but still [JARs](https://docs.oracle.com/javase/tutorial/deployment/jar/basicsindex.html) still are problematic because they contains **date-derived content** (as [ZIP files in general](https://stackoverflow.com/questions/19523063/zip-utility-giving-me-different-md5sum-every-time-in-linux)) which changes hash sum even if code-wise it didn't changed.

You can configure Gradle to not preserve timestamps and then JARs would preserve hash sum between compilations:
```gradle
// build.gradle
allprojects {
    tasks.withType(AbstractArchiveTask).configureEach {
        preserveFileTimestamps = false
        reproducibleFileOrder = true
    }
}
```

You need to wrap it into `allprojects` because if you forgot about that and someday someone will introduce nested module then this module hash will change and it silently break whole caching mechanism of Docker.

**Comment by ljacomet** [comment]
Sorry for the late reply

This feature request is in the backlog of the relevant team and is prioritized by them.

***

The solution might require re-designing the way copy tasks work. They currently create new files.
